contracts/manifold/burnredeem/ERC721BurnRedeem.sol

```diff
@@ -0,0 +1,330 @@
+            _burn(msg.sender, burnToken.contractAddress, burnToken.id);
+
+            groupCounts[burnToken.groupIndex] += 1;
+        }
+
+        for (uint256 i = 0; i < groupCounts.length; i++) {
+            require(groupCounts[i] == _burnRedeem.burnSet[i].requiredCount, "Invalid number of tokens");
+        }
+
+        // Do redeem
+        _mint(creatorContractAddress, index, _burnRedeem, msg.sender);
+    }
+
+    /**
+     * See {ICreatorExtensionTokenURI-tokenURI}.
+     */
+    function tokenURI(address creatorContractAddress, uint256 tokenId) external override view returns(string memory uri) {
+        RedeemToken memory token = _redeemTokens[creatorContractAddress][tokenId];
+        require(token.burnRedeemIndex > 0, "Token does not exist");
+        BurnRedeem memory _burnRedeem = _burnRedeems[creatorContractAddress][token.burnRedeemIndex];
+
+        string memory prefix = "";
+        if (_burnRedeem.storageProtocol == StorageProtocol.ARWEAVE) {
+            prefix = ARWEAVE_PREFIX;
+        } else if (_burnRedeem.storageProtocol == StorageProtocol.IPFS) {
+            prefix = IPFS_PREFIX;
+        }
+        uri = string(abi.encodePacked(prefix, _burnRedeem.location));
+
+        if (!_burnRedeem.identical) {
+            uri = string(abi.encodePacked(uri, "/", uint256(token.mintNumber).toString()));
+        }
+    }
+
+    /**
+     * @dev See {IERC721Receiver-onERC721Received}.
+     */
+    function onERC721Received(
+        address,
+        address from,
+        uint256 id,
+        bytes calldata data
+    ) external override returns(bytes4) {
+        _onERC721Received(from, id, data);
+        return this.onERC721Received.selector;
+    }
+
+    /**
+     * @notice ERC721 token transfer callback
+     * @param from      the person sending the tokens
+     * @param id        the token id of the burn token
+     * @param data      bytes indicating the target burnRedeem and, optionally, a merkle proof that the token is valid
+     */
+    function _onERC721Received(
+        address from,
+        uint256 id,
+        bytes calldata data
+    ) private {
+        // Check calldata is valid
+        require(data.length % 32 == 0, "Invalid data");
+
+        address creatorContractAddress;
+        uint256 burnRedeemIndex;
+        uint256 burnItemIndex;
+        bytes32[] memory merkleProof;
+        (creatorContractAddress, burnRedeemIndex, burnItemIndex, merkleProof) = abi.decode(data, (address, uint256, uint256, bytes32[]));
+
+        BurnRedeem storage _burnRedeem = _burnRedeems[creatorContractAddress][burnRedeemIndex];
+
+        _validateBurnRedeem(_burnRedeem);
+        require(
+            _burnRedeem.burnSet.length == 1 &&
+            _burnRedeem.burnSet[0].requiredCount == 1,
+            "Not a 1:1 burn redeem"
+        );
+
+        // Check that the burn token is valid
+        BurnItem memory burnItem = _burnRedeem.burnSet[0].items[burnItemIndex];
```

Comment:
ah nvm i see this is hardcoded now. this won't work for multiple groups!

Reply:
Should fix this Paul or make it more gas efficient.

Reply:
Clarifying that multi-group burn redeems won't work with the onERC721Received transfer hook (see `require` statement on line 225). We are only receiving one token, so anything configuration that requires more than one token to be burned will need to be redeemed with the `burnRedeem` function. How can do we make it more gas efficient?

Reply:
Misread. Thought it was being called in the set as well but I see you have validateBurnItem for that.

Reply:
Riiiiiiiiiight only 1155 has the batch receive function. Totally forgot about that.